src/pages/index.tsx

```diff
@@ -4,19 +4,108 @@ import { Layout } from "../components/Layout";
+  const [searchActive, SetSearchActive] = useState(false);
+  const [searchCategory, setSearchCategory] = useState<BlogPostCategory[]>([]);
+  const [searchWord, setSearchWord] = useState<string>("");
+  const debounce = useDebounce(searchWord, 500);
   const router = useRouter();
+
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search);
+
+    const searchParam = params.get("search");
+    if (searchParam) {
+      setSearchWord(searchParam);
+    }
+
+    const categoryParam = params.get("category");
+    if (categoryParam !== null) {
+      const validCategories = Object.values(BlogPostCategory);
+      if (validCategories.includes(categoryParam as BlogPostCategory)) {
+        const categoryArray: BlogPostCategory[] = [
+          categoryParam as BlogPostCategory,
+        ];
+        setSearchCategory(categoryArray);
+      }
+    }
+  }, []);
+
+  const POSTS_PER_PAGE = 8;
+
+  const { data, isLoading, error } = api.blogPost.getBlogPosts.useQuery({
+    page: paginationPage,
+    perPage: POSTS_PER_PAGE,
+    categories: searchCategory,
+    search: debounce,
+  });
+  const handlePaginationPageChange = (paginationPage: number) => {
+    setPaginationPage(paginationPage);
+  };
+  const paginationPosts = usePagination(paginationPage, data?.total);
+  const handleSearchCategory = (category: BlogPostCategory[]) => {
+    setSearchCategory(category);
+    history.pushState(
+      null,
+      "search",
+      `?search=${searchWord}&category=${searchCategory}`
+    );
+  };
+  const handleSearchWord = (word: string) => {
+    setSearchWord(word);
+    history.pushState(
+      null,
+      "search",
+      `?search=${searchWord}&category=${searchCategory}`
+    );
+  };
+
   return (
     <Layout>
-      <HStack spacing={4}>
+      <HStack spacing={4} mb={5}>
         <Button onClick={() => router.replace("/login")}>Login</Button>
         <Button onClick={() => router.replace("/register")}>Register</Button>
         <Button onClick={() => router.replace("/user/profile")}>
           User Profile
         </Button>
+        <Button
+          onClick={() => {
+            SetSearchActive(!searchActive);
+            history.pushState(null, "search", `?search`);
+          }}
+        >
+          Search
+        </Button>
       </HStack>
-      <Posts />
+      {searchActive ? (
+        <SearchPost
+          handleSearchCategory={handleSearchCategory}
+          handleSearchWord={handleSearchWord}
+          handlePaginationPageChange={handlePaginationPageChange}
+        />
+      ) : null}
+      <Posts
+        postsQuery={data?.data}
+        paginationPage={paginationPage}
+        postsTotal={data?.total}
+        isLoading={isLoading}
+        error={error}
+      />
+      {data?.total ? (
+        <Pagination
+          paginationPosts={paginationPosts}
+          handlePaginationPageChange={handlePaginationPageChange}
+        />
+      ) : null}
     </Layout>
   );
 };
```

Comment:
Please add a logic to blog list page change - every time blog list page is changed, application should scroll to top. Otherwise, when I go to second page, scroll down and click first page, I'm on the bottom